pkg/scm/git/clone.go

```diff
@@ -19,86 +22,94 @@ type Clone struct {
-		} else {
-			glog.V(1).Infof("Downloading %q ...", config.Source)
-		}
+	ref := "HEAD"
+	if config.Ref != "" {
+		ref = config.Ref
+	}
 
-		// If we have a specific checkout ref, use submodule update instead of recursive
-		// Otherwise the versions will be incorrect.
-		if hasRef && !config.IgnoreSubmodules {
-			glog.V(2).Infof("Cloning sources (deferring submodule init) into %q", targetSourceDir)
-		} else if !hasRef && !config.IgnoreSubmodules {
-			glog.V(2).Infof("Cloning sources and all Git submodules (recursive) into %q", targetSourceDir)
-		} else {
-			glog.V(2).Infof("Cloning sources (ignoring submodules) into %q", targetSourceDir)
-		}
+	if strings.HasPrefix(config.Source, "file://") {
+		s := strings.TrimPrefix(config.Source, "file://")
 
-		if err := c.Clone(config.Source, targetSourceDir, cloneConfig); err != nil {
-			glog.V(1).Infof("Git clone failed: %+v", err)
-			return nil, err
-		}
-
-		if hasRef {
-			if err := c.Checkout(targetSourceDir, config.Ref); err != nil {
-				return nil, err
-			}
-			glog.V(1).Infof("Checked out %q", config.Ref)
-			if !config.IgnoreSubmodules {
-				if err := c.SubmoduleUpdate(targetSourceDir, true, true); err != nil {
-					return nil, err
-				}
-				glog.V(1).Infof("Updated submodules for %q", config.Ref)
-			}
-		}
-
-		if len(config.ContextDir) > 0 {
-			originalTargetDir := filepath.Join(config.WorkingDir, api.Source)
-			c.RemoveDirectory(originalTargetDir)
-			// we want to copy entire dir contents, thus we need to use dir/. construct
-			path := filepath.Join(targetSourceDir, config.ContextDir) + string(filepath.Separator) + "."
-			err := c.Copy(path, originalTargetDir)
+		if util.UsingCygwinGit {
+			var err error
+			s, err = util.ToSlashCygwin(s)
 			if err != nil {
+				glog.V(1).Infof("ToSlashCygwin failed: %+v", err)
 				return nil, err
 			}
-			info = c.GetInfo(targetSourceDir)
-			c.RemoveDirectory(targetSourceDir)
-		} else {
-			info = c.GetInfo(targetSourceDir)
 		}
+		config.Source = "file://" + s
+	}
 
-		if len(config.ContextDir) > 0 {
-			info.ContextDir = config.ContextDir
-		}
+	if len(config.ContextDir) > 0 {
+		targetSourceDir = filepath.Join(config.WorkingDir, api.ContextTmp)
+		glog.V(1).Infof("Downloading %q (%q) ...", config.Source, config.ContextDir)
+	} else {
+		glog.V(1).Infof("Downloading %q ...", config.Source)
+	}
 
-		return info, nil
+	if !config.IgnoreSubmodules {
+		glog.V(2).Infof("Cloning sources into %q", targetSourceDir)
+	} else {
+		glog.V(2).Infof("Cloning sources (ignoring submodules) into %q", targetSourceDir)
 	}
-	// we want to copy entire dir contents, thus we need to use dir/. construct
-	path := filepath.Join(config.Source, config.ContextDir) + string(filepath.Separator) + "."
-	if !c.Exists(path) {
-		return nil, errors.NewSourcePathError(path)
+
+	cloneConfig := api.CloneConfig{Quiet: true}
+	err = c.Clone(config.Source, targetSourceDir, cloneConfig)
+	if err != nil {
+		glog.V(1).Infof("Git clone failed: %+v", err)
```

Comment:
The `+` is just misleading, as it doesn't apply here IIUC. Suggestion: `%s` or `%v`.